Build .rpm packages for linux arm64

wtgodbe opened the pull request:
Part of https://github.com/dotnet/aspnetcore/issues/40253, https://github.com/dotnet/aspnetcore/issues/40252. Test build: https://dev.azure.com/dnceng/internal/_build/results?buildId=1638723&view=results

CC @NikolaMilosavljevic

wtgodbe committed:
Build .rpm packages for linux arm64

wtgodbe committed:
Fixup

dougbu commented on .azure/pipelines/ci.yml:
```diff
@@ -474,17 +474,27 @@ stages:
+            --no-build-nodejs
+            --no-build-java
+            -p:OnlyPackPlatformSpecificPackages=true
+            -p:AssetManifestFileName=aspnetcore-Linux_arm64.xml
```
Official build should break if this setting exists in anything except the last step of the job. You may be :ok: w/o `$(_PublishArgs)` but that just leaves this line as redundant and confusing.

wtgodbe replied:
Good catch - I copied this from the Linux x64 job, so I guess the duplication exists there too. I'll remove it from both.

wtgodbe committed:
Remove dupe asset manifest params

wtgodbe commented:
New build: https://dev.azure.com/dnceng/internal/_build/results?buildId=1638654&view=results

wtgodbe committed:
Don't build native

wtgodbe commented:
https://dev.azure.com/dnceng/internal/_build/results?buildId=1638723&view=results

dougbu reviewed (approved):
Looks fine but I'm not sure why C++ projects are building (and breaking) in the updated build job&hellip;

wtgodbe committed:
Fix RPM build condition

wtgodbe commented:
> Looks fine but I'm not sure why C++ projects are building (and breaking) in the updated build job…

I think https://github.com/dotnet/aspnetcore/pull/40485/commits/ca0c6d28edcaa4da4c393ad2b1fefb055a11638e will fix it. https://dev.azure.com/dnceng/internal/_build/results?buildId=1638815&view=results

wtgodbe committed:
Get binlog

wtgodbe committed:
Fix arch

wtgodbe committed:
Fixup

wtgodbe commented:
Got this mostly working internally, will push to the public branch once it's fully working (and I'm removed the binlog I added): https://dev.azure.com/dnceng/internal/_build/results?buildId=1638974&view=results

wtgodbe committed:
No binlog

wtgodbe commented:
Ok, latest internal builds looks like it worked. @NikolaMilosavljevic how did you test the runtime linux-arm64 installers?

wtgodbe commented:
/backport to release/6.0

dougbu commented:
@NikolaMilosavljevic something looks wrong here. All of our Linux builds produce files named e.g. aspnetcore-targeting-pack-7.0.0-preview.3.22151.16.rpm but we use platform-specific names for the Windows targeting packs. What is the intention here -- aspnetcore-targeting-pack-7.0.0-preview.3.22151.16.rpm or aspnetcore-targeting-pack-7.0.0-preview.3.22151.16-arm64.rpm❔

We'll need to make a change either way -- stop uploading the extra targeting packs (and remove duplicates from the asset manifests) or change the names.

dougbu commented:
Also why bother w/ Mariner-specific targeting packs❔ aspnetcore-targeting-pack-7.0.0-preview.3.22151.16-cm.1.rpm and aspnetcore-targeting-pack-7.0.0-preview.3.22151.16-cm.2.rpm look weird if we're not spelling out the architecture. Those two files and aspnetcore-targeting-pack-7.0.0-preview.3.22151.16.rpm will _never_ have different content.

/fyi we produce just one aspnetcore-targeting-pack-7.0.0-preview.3.22151.16.deb file.

dougbu commented:
If it matters, I don't understand the runtime (really Microsoft.DotNet.SharedFramework.Sdk from Arcade) choices in this area:

![image](https://user-images.githubusercontent.com/6431421/156866686-f34c72ef-9377-48d2-b2b0-af4c22c4bfc4.png)

There's also a dotnet-targeting-pack-7.0.0-preview3-22154-x64.rpm file. Why not just one dotnet-targeting-pack-7.0.0-preview3-22154.rpm file for all❔

Regardless, the RID in the shared Fx .rpm files for ARM64 should probably be `aarch64` based on my reading of a few Microsoft.DotNet.SharedFramework.Sdk.

Thoughts @nikolam @safern❔

dougbu commented:
Bit more info: dotnet/runtime may need arch-specific .rpm files because their PlatformManifest.txt files vary between targeting packs e.g.

``` diff
373,374c373,374
< mscordaccore_amd64_amd64_7.0.22.15502.dll|Microsoft.NETCore.App.Ref|7.0.0.0|7.0.22.15502
< mscordaccore_amd64_amd64_7.0.22.15502.dll|Microsoft.NETCore.App.Ref|7.0.0.0|7.0.22.15502
---
> mscordaccore_amd64_arm64_7.0.22.15502.dll|Microsoft.NETCore.App.Ref|7.0.0.0|7.0.22.15502
> mscordaccore_arm64_arm64_7.0.22.15502.dll|Microsoft.NETCore.App.Ref|7.0.0.0|7.0.22.15502
```

We don't have anything similar in dotnet/aspnetcore though the -cm.1 / -cm.2 requirement (for distinct signing) probably still applies. Depending on the upstream expectations and binary content in the installers themselves (❔❔❔), we may just need to fix the duplication in our asset manifests.

dougbu committed:
Avoid duplicate targeting-pack RPM Build
- if an RPM containing platform-neutral executables is platform-neutral, this should be enough

dougbu commented:
Test build matching this iteration: https://dev.azure.com/dnceng/internal/_build/results?buildId=1647192&view=results

dougbu commented:
To summarize:

1. Is it a bug that our shared Fx RPM is named like aspnetcore-runtime-7.0.0-preview.3.22155.3-**arm64**.rpm❔
2. Do we require arm64-specific targeting pack RPMs❔
3. If (2) is "no", should we remove `--architecture`  from the `fpm` command line when producing the targeting pack RPMs❔
4. I have FR questions open about how to test RPMs on an ARM64 machine. Any suggestions❔

dougbu commented:
/azp run

NikolaMilosavljevic commented:
> To summarize:
> 
> 1. Is it a bug that our shared Fx RPM is named like aspnetcore-runtime-7.0.0-preview.3.22155.3-**arm64**.rpm❔
> 2. Do we require arm64-specific targeting pack RPMs❔
> 3. If (2) is "no", should we remove `--architecture`  from the `fpm` command line when producing the targeting pack RPMs❔
> 4. I have FR questions open about how to test RPMs on an ARM64 machine. Any suggestions❔

1. Package filename does not matter much, but for consistency it would be good if it used the correct arch name `aarch64`.
2. We do need aarch64 targeting pack RPM. Package identity is in package metadata and arch needs to match target platform architecture.

dougbu commented:
@NikolaMilosavljevic as things stand, I can fix this PR and #40504 but am dubious about getting the 6.0.x fix in before the branch closes today due to component governance problems I'm fixing.

I also won't be able to confirm the new RPMs are usable on a Linux ARM64 machine w/o guidance and an appropriate system. I'm not getting very far in my Teams conversation w/ FR.

dougbu committed:
Correct RPM arch name, include in targeting pack

dougbu commented:
Latest test build: https://dev.azure.com/dnceng/internal/_build/results?buildId=1650771&view=results

dougbu committed:
Revert "Avoid duplicate targeting-pack RPM Build"

This reverts commit 77f81b8d696d663249a54c56be4a1d70fe68668f

dougbu commented:
Now: https://dev.azure.com/dnceng/internal/_build/results?buildId=1650784&view=results

dougbu commented:
Example names are now
- [aspnetcore-targeting-pack-7.0.0-preview.3.22157.12-cm.1-aarch64.rpm](https://dev.azure.com/dnceng/_apis/resources/Containers/9779258/BlobArtifacts?itemPath=BlobArtifacts%2Faspnetcore-targeting-pack-7.0.0-preview.3.22157.12-cm.1-aarch64.rpm)
- [aspnetcore-targeting-pack-7.0.0-preview.3.22157.12-cm.1-amd64.rpm](https://dev.azure.com/dnceng/_apis/resources/Containers/9779258/BlobArtifacts?itemPath=BlobArtifacts%2Faspnetcore-targeting-pack-7.0.0-preview.3.22157.12-cm.1-amd64.rpm)

1. Should `amd64` be `x64`❔ That would match the RPMs from dotnet/runtime but I don't want to reason from a sample set of one.
2. Should `-cm.1` or `-cm.2` appear after the architecture instead e.g. aspnetcore-targeting-pack-7.0.0-preview.3.22157.12-amd64-cm.1.rpm❔

dougbu committed:
Rename x64 RPM to "x64"

dougbu commented:
Fixed `x64` names. New test build is https://dev.azure.com/dnceng/internal/_build/results?buildId=1651818&view=results

wtgodbe commented:
Validation confirmed the install works - thanks for the help @dougbu and @NikolaMilosavljevic!
